### PR TITLE
Improvements for REPL output pre-processing and locref navigation 

### PR DESCRIPTION
**Pull request by vspinu**
Improving on #2043 a bit for navigation. Pass through file metatdata when available from location references and no info from var names could be retrieved. `cider-find-file`  does a good job  in those cases.

### CODE REVIEW

File: cider-repl.el [hunk 1]
```diff
@@ -1046,10 +1046,10 @@ regexes from `cider-locref-regexp-alist' to infer locations at point."
   (if-let ((loc (cider-locref-at-point pos)))
       (let* ((var (plist-get loc :var))
              (line (plist-get loc :line))
-             (file (if var
-                       (or (cider-sync-request:ns-path var)
-                           (nrepl-dict-get (cider-sync-request:info var) "file"))
-                     (plist-get loc :file))))
+             (file (or (when var
```
Comment (bbatsov) [inline cider-repl.el]: Who does this improve the code? Seems exactly the same to me.

File: cider-repl.el [hunk 2]
```diff
@@ -561,7 +561,7 @@ When there is a possible unfinished ansi control sequence,
                    (buffer-local-value 'cider-repl--ns-forms-plist connection)
                    ns)))))
 
-(defvar cider-repl--root-ns-highlight-template "\\<\\(%s\\)[^$/: \t\n]+"
+(defvar cider-repl--root-ns-highlight-template "\\<\\(%s\\)[^$/: \t\n()]+"
```
Comment (bbatsov) [inline cider-repl.el]: What's the purpose of this change?

File: cider-repl.el [hunk 3]
```diff
@@ -1051,7 +1051,7 @@ regexes from `cider-locref-regexp-alist' to infer locations at point."
                              (nrepl-dict-get (cider-sync-request:info var) "file")))
                        (plist-get loc :file))))
         (if file
-            (cider--jump-to-loc-from-info (nrepl-dict "file" file "line" line))
+            (cider--jump-to-loc-from-info (nrepl-dict "file" file "line" line) t)
```
Comment (bbatsov) [inline cider-repl.el]: I think it's subjective whether this is better or worse - in such cases I always add a defcustom. 

P.S. Mind the commit message style. 😃

Comment (vspinu) [reply cider-repl.el]: It's not subjective. These pseudo links now behave like links in cider stacktrace buffer. Also like links in every other buffer which has a bunch of links in it (occur, grep, compilation etc) behave like that. It's common to click on multiple error references till you get what you want. 

> Mind the commit message style. :smiley:

What do you mean more concretely?

File: cider-repl.el (hunk 2 above)
Comment (vspinu) [reply cider-repl.el]: Improve original regexp. Currently things like ns1.foo.bar(file.clj:123) are highlighted all the way to 123. This change stops them at (. 

Originally I thought all those reference have $ in their name, but that's not always the case.

File: cider-repl.el (hunk 3 above)
Comment (xiongtx) [reply cider-repl.el]: > What do you mean more concretely?

Follow [this guide](https://chris.beams.io/posts/git-commit/). 

> When clicked on locref in REPL, jump to other window.

doesn't adhere to #5.

File: cider-repl.el (hunk 1 above)
Comment (vspinu) [reply cider-repl.el]: The difference is that now `:file` metadata is picked up when first two resolutions fail (those within or). This happens when a var is a protocol method. This was also the original intention btw.

File: cider-repl.el (hunk 3 above)
Comment (vspinu) [reply cider-repl.el]: Ohh, dot at the end of the message. Fixing that :) The subject line is already imperative mood.

Comment (xiongtx) [reply cider-repl.el]: > The subject line is already in imperative mood.

I'm not a stickler for this stuff, but no, it's not. 

> Jump to other window when clicking on locref in REPL

would be.

File: cider-repl.el (hunk 1 above)
Comment (bbatsov) [reply cider-repl.el]: Ah, I see your point now - `var` might be present, but its branch might return `nil`.  Add some comments as it's somewhat hard to follow the what the assumptions here otherwise.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 ### Bugs Fixed
 
+* Improve file retrieval when jumping to REPLs location references and Clojure vars cannot be resolved.
```
Comment (bbatsov) [inline CHANGELOG.md]: How come you can just to something that's not resolved? I guess you made some mistake in the description here.

File: cider-repl.el (hunk 1 above)
Comment (vspinu) [reply cider-repl.el]: Yes, I used "var" loosely to refer to the variable at point, which need not correspond to clojure's var. Will add a couple of notes.

File: CHANGELOG.md (hunk 4 above)
Comment (vspinu) [reply CHANGELOG.md]: That code is recent. I added it in  #2043 and now I am just improving on it. 

I just said what I did there - when locrefs cannot be referenced from variables at point, use file metadata from the locrefs themselves. I can state it more generically "Improve locations lookup by using file metadata when available". Not accurate but conveys the improvement. 

AFAIC, this need not be documented at all as being a small and technical fix on a recent change, but knowing how zealous you are on change-logs I added it.

**Comment by vspinu** [comment]
I am piling a bunch of related changes and improvements here. With a bit more doc this time, so should be self explanatory. 

Also fixing failing tests to account for recent buttercup changes.

Two notable new features:

 - `cider-preoutput-hook` which implements a customizable pipeline of output pre-processing and includes most of our output pre-processing as "middleware". 
 - `cider-repl-highlight-spec-keywords` pre-output processor which highlights clojure.spec "explain" keywords. Clojure spec explain output is extremely difficult to read.. Make it a bit easy by showing this:

spec-asert2
  
instead of

spec-asert1

**Comment by xiongtx** [comment]
> I am piling a bunch of related changes and improvements here. 

It'd be better to open another PR. A bunch of unrelated changes in a single PR is hard to review.

File: cider-repl.el (hunk 3 above)
Comment (bbatsov) [reply cider-repl.el]: Indeed. The commit messages should ideally always start with a verb.
